### ISSUE TITLE
all_to_all with axis_index_group argument broken

### ISSUE THREAD

**Issue by mganahl** [jax-ml/jax#5861]
Hi! I'm hitting some unexpected behaviour when using `all_to_all`. I'm trying to do partial swaps of a pmapped axis. The following code performes the parallel version of 
`np.arange(8).reshape(2,2,2).transpose((0,2,1)).reshape(4,2)`, and works as expected:

```python

import os
os.environ['XLA_FLAGS']='--xla_force_host_platform_device_count=4'
import jax

from jax import pmap
import numpy as np

split_axis=0
concat_axis=0
res = pmap(lambda x: jax.lax.all_to_all(x, 'i',
                                  split_axis, 
                                  concat_axis,
                                  axis_index_groups=[[0,1],[2,3]]), 
           axis_name='i')(np.arange(8).reshape(4,2))
print('actual')
print(np.array(res))
#[[0 2]
#  [1 3]
#  [4 6]
#  [5 7]]
print('expected')
print(np.arange(8).reshape(2,2,2).transpose((0,2,1)).reshape(4,2))
#[[0 2]
#  [1 3]
#  [4 6]
#  [5 7]]
```
I believe the following code _should_ perform the parallel version of 
`np.arange(8).reshape(2,2,2).transpose((2,1,0)).reshape(4,2)`,
```python 
split_axis=0
concat_axis=0
res = pmap(lambda x: jax.lax.all_to_all(x, 'i',
                                        split_axis, 
                                        concat_axis,
                                        axis_index_groups=[[0,2],[1,3]]), 
           axis_name='i')(np.arange(8).reshape(4,2))

print('actual')
print(np.array(res))
#[[0 4]
# [3 7]
# [0 4]
# [3 7]]
print('expected')
print(np.arange(8).reshape(2,2,2).transpose((2,1,0)).reshape(4,2))
#[[0 4]
# [2 6]
# [1 5]
# [3 7]]
```
alas, it doesn't! The actual result of the second snippet is somewhat surprising to me. Maybe I'm just not using `all_to_all` properly.

A side note: The same code, when run on a v3-8, actually hangs, as does `pswapaxes` for similar uses cases (`pmap`ing other functions works as expected).

Any help would be greatly appreciated!
thanks!

edit: this could be related to #5757

edit 2: This seems to be a driver issue. On a v3-8 this works with tpu_driver_nightly, but not other drivers.

**Comment by mhauru** [comment]
The above behavior seems to be a CPU specialty, on a TPU v2 donut it works fine. Should this perhaps be considered an XLA bug?

**Comment by mganahl** [comment]
@jacksonwb

**Comment by mganahl** [comment]
@skye Just a quick ping on the status of this!

**Comment by skye** [comment]
Hi, sorry for the delay on this! I filed an internal XLA bug (b/187117499 for Googlers) and will keep this thread updated.

**Comment by rajasekharporeddy** [comment]
Hi @mganahl 

It looks like this has been resolved. I executed the mentioned code with JAX version 0.4.23 on Colab CPU and JAX version 0.3.25 on colab TPU. Now the second code mentioned also produced the output as expected.

```python
import os
os.environ['XLA_FLAGS']='--xla_force_host_platform_device_count=4'
import jax

from jax import pmap
import numpy as np


split_axis=0
concat_axis=0
res = pmap(lambda x: jax.lax.all_to_all(x, 'i',
                                        split_axis, 
                                        concat_axis,
                                        axis_index_groups=[[0,2],[1,3]]), 
           axis_name='i')(np.arange(8).reshape(4,2))

print('actual')
print(np.array(res))

print('expected')
print(np.arange(8).reshape(2,2,2).transpose((2,1,0)).reshape(4,2))
```
Output:
```
actual
[[0 2]
 [1 3]
 [4 6]
 [5 7]]
expected
[[0 2]
 [1 3]
 [4 6]
 [5 7]]
```

Kindly find the gists on [CPU](https://colab.research.google.com/gist/rajasekharporeddy/f29116ca9f3b5c89fd28e34008e8000d/-cpu_5861.ipynb) and [TPU](https://colab.research.google.com/gist/rajasekharporeddy/5ce1d454760c8165b3948ecbed6734dd/-tpu_5861.ipynb) for reference.

Thank you
